Add e2e tests for multiAttach

mkimuram opened the pull request:
**What type of PR is this?**
 /kind feature

**What this PR does / why we need it**:
This PR improves test coverage for raw block feature.
The background for this improvement is described in #74545.
Reconstruction tests required in #74545 will be handled in another PR.

**Which issue(s) this PR fixes**:


**Special notes for your reviewer**:
/sig storage
cc @bswartz @wongma7 @msau42 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

mkimuram committed:
Add e2e tests for multiAttach

msau42 commented:
/ok-to-test

mkimuram commented:
@msau42 @wongma7 

First commit doesn't include "Reconstruction of block devices" and only focus on mulit-attach cases. 
I would appreciate it if you could start review for this first, while I'm working on implementing reconstruction test codes.

Note that the relationships between the list of planned tests and the test cases in the code are as follows: 

- Multiple pods requesting same block volume
Below test case should cover this:
"should attach the same volume with the same volume mode to the different pod on the same node"
(This also tests filesystem case, which can be skipped later if not necessary.)

- Single pod requests both filesystem and raw block volumes for the same volume
Below test case should cover this:
"should attach two volumes with different volume mode to the same pod"
Also, the same volume case for both filesystem and block are covered by below case:
"should attach two volumes with the same volume mode to the same pod"

- Multiple pods: one requests filesystem and one requests raw block for the same volume
It seems that  ```volumeMode``` is per PVC and mapping for PVC and PV is 1:1,
therefore we might not be able to attach different volumeMode to multiple pod.
If I missed something, please point me out.

wongma7 commented:
for "Multiple pods: one requests filesystem and one requests raw block for the same volume" we'll need to create a 2nd PV with the same `spec`

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,272 @@
+				testpatterns.FsVolModeDynamicPV,
+				testpatterns.BlockVolModeDynamicPV,
+				// Currently, multiple volumes are not generally available for pre-provisoined volume,
+				// because containerized storage servers, such as iSCSI and rbd, are just returning
```
I'm going to make it so the iscsi image creates 2 luns instead of 1. otherwise we cannot test the case where 2 vols are different except for the lun, I want to add that testcase here

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,271 @@
+		l.resource2.pv, err = l.cs.CoreV1().PersistentVolumes().Get(l.resource2.pvc.Spec.VolumeName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Creating pod with the two volumes with the same volume mode")
```
the volume modes might be different

mkimuram replied:
Thank you for pointing it out! I just copy & past when refactoring locally.

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,271 @@
+
+		// Decide the node to deploy both 1st pod and 2nd pod.
+		// In this test, pods are depoyed to the node other than the node
+		// that is defined in l.config.ClientNodeName.
```
why can't it be ClientNodeName?

mkimuram replied:
This intends to deploy both 1st pod and 2nd pod to the same node.

I think that some drivers can properly attach volumes to different nodes at the same time, but others can't.

For example, NFS and CephFS should be able to attach a volume to multiple nodes at the same time and the volume could be consumed by any pods at the same time. However, iSCSI and rbd won't be able to attach a volume to multiple nodes at the same time, especially when it is used as filesystem volume. Because, it will destroy metadata on the filesystem.
(Also, some cloud providers won't allow attaching a volume to multiple nodes, at the same time. And local volume won't be attach to another nodes.)

So, I think that we need to separate tests for the same node case and different node case. And the background of adding this test case was happened in the same node case, I focused this case here. (We could also add different node case.)

mkimuram replied:
I started to think that above is just for the case if ClientNodeName hasn't been set, already. If driver requires to deploy on somewhere for some reason, like hostpath driver do, we should use this value.(In that case, we will still be able to make both pod deploy on the same node, without updating the node again here.)

wongma7 replied:
OK i see, thanks. I don't think a different node case is needed, maybe later. Yeah, probably ClientNodeName is empty sometimes.

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,371 @@
+		defer cleanup()
+
+		l.resource2.pvc = l.resource1.pvc
+		testAttachOneVolumeToTwoPods(l, true /* sameNode */)
```
the last 2 tests will work with preprovisionedPVTestDrivers right? is there value in testing them, besides iscsi? If so, we can wrap the above two in "if is a DynamicPVTestDriver". However, then the testsuite accepted patterns and the preconditions will also need to be changed.

mkimuram replied:
Good point. I will add codes to skip pre-provisioned volume for tests that require two volumes, then add pre-provisioned volume to the list of test patterns to execute. Then, what we originally would like to test can be tested. 

> is there value in testing them, besides iscsi?

I guess that this issue won't be iscsi specific, so testing it for other drivers will have benefit to all drivers.

mkimuram commented:
@wongma7 @msau42 

>for "Multiple pods: one requests filesystem and one requests raw block for the same volume" we'll need to create a 2nd PV with the same spec

I've added below tests by above way that @wongma7 suggested.

```
"should attach the same volume with different volume mode to the different pod on the same node"
"should attach the same volume with different volume mode to the different pod on different node" 
```

They are tested for both block volume case and filesystem volume case.
Also, they are tested for both dynamic provision case and pre-provisioned PV case.




Strangely, test results show that only block volume with dynamic provision cases fails in deletion of first volume. (See below details for tested combinations and results. Especially, it is strange that case 3 passes in this situation.) Are there any reason that you can think of that only this combination fails?

Also, are there are a way to access to detailed logs for the tests to debug more?
(GCE pd is the only driver that I can test for "Block volume with dynamic provision" case, so it would be nice if I could get more logs for them to debug more.)

<details>

 Test | volume1 | volume2 | result
---|--- |---  |---
 1| Dynamic Block | Preprovisioned Block  | Success
 2| Dynamic Block | Preprovisioned File  | Fail
 3| Dynamic File | Preprovisioned Block  | Success
 4| Dynamic File | Preprovisioned File  | Success
 5| Preprovisioned Block | Preprovisioned Block  | Success
 6| Preprovisioned Block | Preprovisioned File  | Success
 7| Preprovisioned File | Preprovisioned Block  | Success
 8| Preprovisioned File | Preprovisioned File  | Success

</details>

wongma7 commented:
go to artifacts http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/74693/pull-kubernetes-e2e-gce/70987/ from the "Job View", e.g. here are some kubelet logs http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/74693/pull-kubernetes-e2e-gce/70987/artifacts/e2e-70987-674b9-minion-group-bkpw/

In this case,

[sig-storage]  In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (block  volmode)] multiAttach should attach the same volume with different  volume mode to the different pod on the same node expand_less | 6m39s
-- | --
test/e2e/storage/testsuites/multiattach.go:297 Persistent Volume pvc-8b06e1f5-3c96-11e9-9f70-42010a280002 not deleted by dynamic provisioner Expected error:     <*errors.errorString \| 0xc0026be3f0>: {         s: "PersistentVolume pvc-8b06e1f5-3c96-11e9-9f70-42010a280002 still exists within 5m0s",     }     PersistentVolume pvc-8b06e1f5-3c96-11e9-9f70-42010a280002 still exists within 5m0s not to have occurred test/e2e/storage/testsuites/base.go:244

UnmountDevice fails repeatedly. Here are some interesting lines:

...
W0302 02:56:10.409869    1644 volume_path_handler.go:126] Warning: Unmap skipped because symlink does not exist on the path: /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/e2e-70987-674b9-dynami-pvc-8b06e1f5-3c96-11e9-9f70-42010a280002/91ca95d2-3c96-11e9-9f70-42010a280002

I0302 02:56:10.409888    1644 operation_generator.go:1094] UnmapVolume succeeded for volume "kubernetes.io/gce-pd/e2e-70987-674b9-dynami-pvc-8b06e1f5-3c96-11e9-9f70-42010a280002" (OuterVolumeSpecName: "volume1") pod "91ca95d2-3c96-11e9-9f70-42010a280002" (UID: "91ca95d2-3c96-11e9-9f70-42010a280002"). InnerVolumeSpecName "pvc-8b06e1f5-3c96-11e9-9f70-42010a280002". PluginName "kubernetes.io/gce-pd", VolumeGidValue ""

E0302 02:56:10.593163    1644 nestedpendingoperations.go:267] Operation for "\"kubernetes.io/gce-pd/e2e-70987-674b9-dynami-pvc-8b06e1f5-3c96-11e9-9f70-42010a280002\"" failed. No retries permitted until 2019-03-02 02:56:11.093118038 +0000 UTC m=+807.121365988 (durationBeforeRetry 500ms). Error: "UnmapDevice.RemoveMapPath failed for volume \"pvc-8b06e1f5-3c96-11e9-9f70-42010a280002\" (UniqueName: \"kubernetes.io/gce-pd/e2e-70987-674b9-dynami-pvc-8b06e1f5-3c96-11e9-9f70-42010a280002\") on node \"e2e-70987-674b9-minion-group-bkpw\" : device or resource busy"
...

pod "91ca95d2-3c96-11e9-9f70-42010a280002" is "pod1"  security-context-91c9db69-3c96-11e9-82d0-1a0b1b15e36b with volumeMapPath "/var/lib/kubelet/pods/91ca95d2-3c96-11e9-9f70-42010a280002/volumeDevices/kubernetes.io~gce-pd"

I do not see any attempts to unmount pod2's mount.

Maybe we found a bug already?

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,437 @@
+		By("Checking if read/write to the volume in 2nd pod works properly")
+		utils.CheckReadWriteToPath(pod2, l.resource2.pattern.VolMode, "/mnt/volume1")
+
+		// Delete pod1
```
pod2

wongma7 commented:
2 | Dynamic Block | Preprovisioned File | Fail

So pod2 using "cloned-pv-p8zks" is deleted first, but it is not removed from desired state of world so it is not unmounted. Then pod1 using "pvc-8b06e1f5-3c96-11e9-9f70-42010a280002" is deleted and removed from desired state of world 

I0302 02:56:10.372199    1644 desired_state_of_world_populator.go:265] Removing volume from desired state for volume "pvc-8b06e1f5-3c96-11e9-9f70-42010a280002" (UniqueName: "kubernetes.io/gce-pd/e2e-70987-674b9-dynami-pvc-8b06e1f5-3c96-11e9-9f70-42010a280002") pod "security-context-91c9db69-3c96-11e9-82d0-1a0b1b15e36b" (UID: "91ca95d2-3c96-11e9-9f70-42010a280002")

UnmountDevice for pod1 fails since pod2's volume was never unmounted.

On the other hand:

3 | Dynamic File | Preprovisioned Block | Success

This succeeds because it doesn't matter if pod2's volume was unmapped, UnmountDevice for pod1 will succeed since it will UnmountVolume first.

I am (re)-learning how desired state of world works, but I am sure there is a bug there.

mkimuram commented:
@wongma7 

Thank you for quickly sharing your observation. I'll also check the codes around them.

wongma7 commented:
@mkimuram so we can't use 2 pvs to represent same volume after all, my bad...

unique volume name includes the volume's name (so clone* vs pvc*)
https://github.com/kubernetes/kubernetes/blob/9fd23b217800a312a45dbaedb70f4ffa65ec661b/pkg/volume/util/util.go#L340

if that's the case, then IMO we don't need to test it as it's not supported by kube

mkimuram commented:
@wongma7 

OK
I will make another commit to remove multiple volumeMode to one volume case.

wongma7 commented:
okay. I have saved this revision locally so I can reuse most of the code for iscsi anyway, will ping you when that is ready for review (after this merges, probably).

mkimuram commented:
@wongma7 

I removed the tests and quashed the commit. PTAL

If we prefer to merge this PR without reconstruction tests and go ahead with your work on top of this.
I think that we'd better remove "Which issue(s) this PR fixes:" in my first comment, before merge.
(And maybe also change subject to "Add e2e tests for multiAttach")

wongma7 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,323 @@
+		}
+		index := rand.Intn(len(nodes.Items))
+		node1 = nodes.Items[index].Name
+		node2 = nodes.Items[index%len(nodes.Items)].Name
```
I don't think this will always work, e.g. 1%2=1

mkimuram replied:
Oh, I meant ```(index + 1)%len(nodes.Items)```, and typed differently. Thank you for noticing it.

mkimuram replied:
@wongma7 

Fixed and squashed the commit. PTAL

wongma7 commented:
agree, let's leave #74545 open and decide how to proceed for the other cases. Reconstruction will be disruptive. And multi-pod same-volume, I see now we can't test

wongma7 commented:
/lgtm
thanks!

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -47,6 +47,7 @@ var csiTestSuites = []func() testsuites.TestSuite{
 	testsuites.InitSubPathTestSuite,
 	testsuites.InitProvisioningTestSuite,
 	testsuites.InitSnapshottableTestSuite,
+	testsuites.InitMultiAttachTestSuite,
```
nit: let's not call it multi attach. That will confuse it with RWX drivers.  Maybe multi mount? multi volumes?

msau42 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,323 @@
+	// [   node1   ]
+	//   \      /     <- same volume mode
+	//   [volume1]
+	It("should attach the same volume with the same volume mode to the different pod on the same node", func() {
```
I think this test case is already handled here: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/storage/testsuites/provisioning.go#L229

msau42 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,323 @@
+	// [   node1   ] [   node2   ]
+	//         \      /     <- same volume mode
+	//         [volume1]
+	It("should attach the same volume with the same volume mode to the different pod on different node", func() {
```
This test case is going to fail for RWO volumes. This test should be gated by a RWX volume capability.

msau42 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,323 @@
+			nodes := framework.GetReadySchedulableNodesOrDie(cs)
+			nodeName = nodes.Items[rand.Intn(len(nodes.Items))].Name
+		}
+		node1 = nodeName
```
instead of forcing nodes by setting pod.spec.nodename, I would prefer using PodAffinity/Anti-Affinity so that the pod still goes through the scheduler.  You can probably reuse methods that were implemented in the provisioning suite.

msau42 commented on test/e2e/storage/testsuites/multiattach.go:
```diff
@@ -0,0 +1,323 @@
+		l.resource2.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.resource2.pvc)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.resource2.pvc.Namespace, l.resource2.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
```
Instead of waiting for PVCs to be bound before creating the pod, I would prefer that you start the Pod anyway and check for PVC bound at the end.  That way this test case will work with delayed binding too.

mkimuram commented:
@msau42 

The gce tests started to fail after I fixed to deploy pods to different nodes, properly.

Does GCE allow one volume to attach to different nodes at the same time?

If not, I think that there are two choices:
1. Introduce a flag like ```shareable``` and run these tests only when this flag is set (or test that it fails),
2. Remove "different nodes" tests, for now.

Could you let me know which one you prefer?

msau42 commented:
I suggested [here](https://github.com/kubernetes/kubernetes/pull/74693#discussion_r262258907) to add a new driver capability for RWX

bswartz commented:
Guys you know that the hostpath driver fundamentally can't support more than 1 node right? Are you talking about using a different driver for these multi-node tests?

msau42 commented:
Hostpath driver can still support the multi-pods on a single node test.  Or two volumes on the same node.

bswartz commented:
Yeah but even if you use RWO volumes with one pod per volume, the moment you expand the cluster to 2 nodes, the hostpath driver will break because it only creates the volumes on the node where the provisioner sidecar happens to be running.

msau42 commented:
The hostpath driver in the e2e tests is already configured to force the scheduling to a single node

mkimuram commented:
@msau42 

I renamed the test, introduced RWX flag, and modified to use affinity. PTAL
Also, could you let me know your preference whether we should skip duplicated tests or which one to be merge?

msau42 commented on test/e2e/storage/testsuites/multivolume.go:
```diff
@@ -0,0 +1,312 @@
+	//          [pod1]
+	//          /    \      <- same volume mode
+	//   [volume1]  [volume2]
+	It("should attach two volumes with the same volume mode to the same pod", func() {
```
There's already a test case that does this [here](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/storage/persistent_volumes.go#L309) and more generically (can handle more than 2). Maybe that should be adapted to the into the test suite.

msau42 commented on test/e2e/storage/testsuites/multivolume.go:
```diff
@@ -0,0 +1,312 @@
+	//          [pod1]
+	//          /    \      <- different volume mode (only <block, filesystem> pattern is tested)
+	//   [volume1]  [volume2]
+	It("should attach two volumes with different volume mode to the same pod", func() {
```
in the test case names maybe replace "attach" with "access"? to not confuse with the actual attach operation

msau42 commented:
I think we should consolidate duplicate test cases into one.  We can maybe move the test from the provisioning suite over here since that test case isn't really testing provisioning.

mkimuram commented:
@msau42 

Thank you for your quick feedback. I will work on it.

mkimuram commented:
Completed refactoring to make tests usable from other tests, but still working on to consolidate. 
I will continue working on this tomorrow.

mkimuram commented:
@msau42 

I think that I have fixed as suggested. Could you check the code that I preserved the intention of existing code?

mkimuram commented:
/retest

msau42 commented on test/e2e/storage/testsuites/multivolume.go:
```diff
@@ -0,0 +1,468 @@
+// and recheck access to the volumes from the recreated pod
+func TestAccessMultipleVolumesAcrossPodRecreation(f *framework.Framework, cs clientset.Interface, ns string,
+	node framework.NodeSelection, pvcs []*v1.PersistentVolumeClaim, requiresSameNode bool) {
+	By(fmt.Sprintf("Creating pod on %+v with multiple volumes", node))
```
L323 - L340 are repeated. Can we put those into a function?

mkimuram replied:
@msau42 

Thank you for good advice, I fixed as suggested.
Also, I found that read test should be done before write test after pod recreation, but previous code doesn't do it properly.
So, I fixed it to work properly.

PTAL

msau42 commented on test/e2e/storage/testsuites/multivolume.go:
```diff
@@ -0,0 +1,468 @@
+			framework.Skipf("This test doesn't work with pre-provisioned volume -- skipping")
+		}
+
+		init()
```
every test case is doing init/cleanup.  Can this go in Before/AfterEach instead?

mkimuram replied:
Current code is a bit redundant, but it give us a chance to skip tests before driver initialization, per test basis. (Some test can be skipped before the information of driver, but others won't.)
If we prioritize readability over performance, we will be able to choose moving it to Before/AfterEach.

msau42 replied:
I guess for now it's fine, but I eventually want to move to a model where we bring up the driver during test cluster initialization instead of during the test case.

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

mkimuram commented:
/retest

mkimuram commented:
/retest

msau42 commented:
/lgtm
/approve
/kind cleanup
/priority important-soon

Thanks for helping improve our test coverage!

msau42 commented:
/skip

wongma7 commented on test/e2e/storage/testsuites/driveroperations.go:
```diff
@@ -71,8 +72,8 @@ func GetStorageClass(
-			// Name must be unique, so let's base it on namespace name
-			Name: ns + "-" + suffix,
+			// Name must be unique, so let's base it on namespace name and random string
+			Name: ns + "-" + rand.String(4) + "-" + suffix,
```
prefer  `GenerateName`

mkimuram replied:
@wongma7 @msau42 

Changed codes as below and squashed the commit, now sc is named like below:

PTAL

- sc naming
```
STEP: creating a StorageClass multivolume-4412-nfs-scndbcj 
```

- code change
<details>

```
# git diff
diff --git a/test/e2e/storage/testsuites/driveroperations.go b/test/e2e/storage/testsuites/driveroperations.go
index c0e30c3979..4d4298e61a 100644
--- a/test/e2e/storage/testsuites/driveroperations.go
+++ b/test/e2e/storage/testsuites/driveroperations.go
@@ -22,7 +22,7 @@ import (
        storagev1 "k8s.io/api/storage/v1"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-       "k8s.io/apimachinery/pkg/util/rand"
+       "k8s.io/apiserver/pkg/storage/names"
        "k8s.io/kubernetes/test/e2e/framework"
        "k8s.io/kubernetes/test/e2e/storage/testpatterns"
 )
@@ -72,8 +72,8 @@ func GetStorageClass(
                        Kind: "StorageClass",
                },
                ObjectMeta: metav1.ObjectMeta{
-                       // Name must be unique, so let's base it on namespace name and random string
-                       Name: ns + "-" + rand.String(4) + "-" + suffix,
+                       // Name must be unique, so let's base it on namespace name and use GenerateName
+                       Name: names.SimpleNameGenerator.GenerateName(ns + "-" + suffix),
                },
                Provisioner:       provisioner,
                Parameters:        parameters,
```
</details>

wongma7 commented on test/e2e/storage/utils/utils.go:
```diff
@@ -483,3 +487,43 @@ func CheckReadWriteToPath(pod *v1.Pod, volMode v1.PersistentVolumeMode, path str
 	}
 }
+
+func genBinDataFromSeed(len int, seed int64) []byte {
```
could we generate a string uuid instead? https://github.com/kubernetes/apimachinery/blob/ed135c5b96450fd24e5e981c708114fbbd950697/pkg/util/uuid/uuid.go#L30 it will be a bit easier to follow, don't need to worry about encoding then decoding it in the Write function

wongma7 replied:
nvm, there is no way to generate one from a seed

msau42 commented:
/hold
to address @wongma7 's comments

wongma7 commented:
/hold cancel
/lgtm

thank you

msau42 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

msau42 commented:
/retest

wongma7 commented:
@mkimuram , got blocked by a change to test/e2e/storage/csi_mock_volume.go , needs a small rebase

mkimuram commented:
@wongma7 

Sorry for late response. I've rebased and squashed the commit. PTAL

wongma7 commented:
/lgtm

wongma7 commented on test/e2e/framework/pv_util.go:
```diff
@@ -901,6 +911,44 @@ func CreateSecPodWithNodeName(client clientset.Interface, namespace string, pvcl
+			// However, "name", "Name", "ObjectMeta.Name" all got rejected with "not a valid field selector key".
+
+			MatchExpressions: []v1.NodeSelectorRequirement{
+				{Key: "kubernetes.io/hostname", Operator: operator, Values: []string{nodeName}},
```
what happens if nodeName is ""? like if l.config.ClientNodeName is blank

msau42 replied:
I would expect it to fail and not match anything. I'm surprised the tests are passing though?

msau42 replied:
/hold 
to test this out

mkimuram replied:
```SetNodeAffinityRequirement``` is currently only used in a way like:
1. Deploy first pod,
2. Check the actual node name where first pod was deployed,
3. Create node affinity/anti-affinity for the node 

So, currently it is not used for the case that ```nodeName``` is blank, as a result, this bug hasn't been appeared in actual test.

Let me add codes in ```SetNodeAffinityRequirement``` to fail for ```nodeName``` blank case.

mkimuram replied:
I started to think that we don't need to add codes to force failure in this case, because this case should fail properly without this check. Do you agree or are there any actions needed for this?

msau42 replied:
I think this is fine

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

msau42 commented:
/hold cancel

neolit123 commented:
hello, this PR caused `pull-kubernetes-e2e-gce` to flake consistently:
https://prow.k8s.io/job-history/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce

i will log a ticket in k/k.
